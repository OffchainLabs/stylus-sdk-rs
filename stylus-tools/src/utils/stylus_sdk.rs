// Copyright 2025, Offchain Labs, Inc.
// For licensing, see https://github.com/OffchainLabs/stylus-sdk-rs/blob/main/licenses/COPYRIGHT.md

//! Utilities relating to the stylus-sdk itself.

use std::iter;

use alloy::primitives::map::HashMap;
use cargo_metadata::{Metadata, MetadataCommand, Package};
use cargo_util_schemas::manifest::PackageName;

static CONTRACT_DEPENDENCIES: &[&str] = &["alloy-primitives", "alloy-sol-types"];

/// Required dependencies for stylus contracts.
///
/// These dependencies will use the version requirements specified for the stylus-sdk crate. This
/// is important to ensure sdk compatibility when those types are used within the contract.
pub fn contract_dependencies() -> Result<HashMap<String, String>, cargo_metadata::Error> {
    let deps = stylus_sdk_dependencies()?
        .filter(|(name, _req)| CONTRACT_DEPENDENCIES.contains(&name.as_str()))
        .chain(iter::once((
            "stylus-sdk".to_string(),
            env!("CARGO_PKG_VERSION").to_string(),
        )))
        .collect();
    Ok(deps)
}

/// Get all dependencies for stylus-sdk.
///
/// We can use these to determine required dependencies for contracts using the
/// [contract_dependencies] function.
fn stylus_sdk_dependencies() -> Result<impl Iterator<Item = (String, String)>, cargo_metadata::Error>
{
    let package = stylus_sdk_metadata()?;
    let deps = package
        .dependencies
        .into_iter()
        .map(|dep| (dep.name, dep.req.to_string()));
    Ok(deps)
}

/// Get the package metadata for the stylus-sdk crate.
fn stylus_sdk_metadata() -> Result<Package, cargo_metadata::Error> {
    let workspace = workspace_metadata()?;
    let package_name = PackageName::new("stylus-sdk".to_string()).expect("PackageName validation");
    let package = workspace
        .packages
        .into_iter()
        .find(|p| p.name == package_name)
        .expect("Finding stylus-sdk package");
    Ok(package)
}

/// Get metadata for the stylus-sdk-rs workspace.
fn workspace_metadata() -> Result<Metadata, cargo_metadata::Error> {
    let metadata = MetadataCommand::new()
        .current_dir(env!("CARGO_MANIFEST_DIR"))
        .exec()?;
    Ok(metadata)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_contract_dependencies() {
        let deps = contract_dependencies().unwrap();
        assert!(deps["alloy-primitives"].starts_with("^1"));
        assert!(deps["alloy-sol-types"].starts_with("^1"));
        assert_eq!(deps["stylus-sdk"], env!("CARGO_PKG_VERSION"));
    }
}
