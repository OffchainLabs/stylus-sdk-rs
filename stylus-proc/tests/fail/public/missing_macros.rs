// Copyright 2025, Offchain Labs, Inc.
// For licensing, see https://github.com/OffchainLabs/stylus-sdk-rs/blob/main/licenses/COPYRIGHT.md

extern crate alloc;

use stylus_sdk::{
    alloy_primitives::{Address, U256},
    prelude::*,
};

#[storage]
#[entrypoint]
pub struct Contract {}

#[public]
#[implements(Trait1)]
impl Contract {}

// Missing #[public] attribute on trait methods
pub trait Trait1 {
    fn trait1_fn1(&self, _input: U256) -> U256;
}

#[public]
impl Trait1 for Contract {
    fn trait1_fn1(&self, _input: U256) -> U256 {
        todo!()
    }
}

#[public]
pub trait Trait2 {
    fn trait2_fn1(&self, _input: U256) -> U256;
    fn trait2_fn2(&mut self, _to: Address, _value: U256) -> bool;
}

#[public]
impl Trait2 for Contract {
    fn trait2_fn1(&self, _input: U256) -> U256 {
        todo!()
    }

    fn trait2_fn2(&mut self, _to: Address, _value: U256) -> bool {
        todo!()
    }
}

#[public]
pub trait Trait3 {
    fn trait3_fn1(&self, _input: U256) -> U256;
}

// Missing #[public] attribute on impl block
impl Trait3 for Contract {
    fn trait3_fn1(&self, _input: U256) -> U256 {
        todo!()
    }
}

fn main() {}
