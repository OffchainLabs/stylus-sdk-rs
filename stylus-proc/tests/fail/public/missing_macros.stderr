error[E0407]: method `__stylus_trait_and_impl_must_be_tagged_with_public_macro` is not a member of trait `Trait1`
  --> tests/fail/public/missing_macros.rs:24:1
   |
24 | #[public]
   | ^^^^^^^^^ not a member of trait `Trait1`
   |
   = note: this error originates in the attribute macro `public` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0046]: not all trait items implemented, missing: `__stylus_trait_and_impl_must_be_tagged_with_public_macro`
  --> tests/fail/public/missing_macros.rs:54:1
   |
48 | #[public]
   | --------- `__stylus_trait_and_impl_must_be_tagged_with_public_macro` from trait
...
54 | impl Trait3 for Contract {
   | ^^^^^^^^^^^^^^^^^^^^^^^^ missing `__stylus_trait_and_impl_must_be_tagged_with_public_macro` in implementation
